crates/biome_js_analyze/src/lint/nursery/no_shadow.rs

```diff
@@ -0,0 +1,167 @@
+use biome_analyze::{Rule, RuleDiagnostic, RuleSource, context::RuleContext, declare_lint_rule};
+use biome_console::markup;
+use biome_js_semantic::Binding;
+use biome_js_syntax::{
+    JsClassExpression, JsFunctionExpression, JsIdentifierBinding, JsVariableDeclarator,
+};
+use biome_rowan::{AstNode, SyntaxNodeCast, TokenText, declare_node_union};
+
+use crate::services::semantic::Semantic;
+
+declare_lint_rule! {
+    /// Disallow variable declarations from shadowing variables declared in the outer scope.
+    ///
+    /// Shadowing is the process by which a local variable shares the same name as a variable in its containing scope. This can cause confusion while reading the code and make it impossible to access the global variable.
+    ///
+    /// See also: [`noShadowRestrictedNames`](http://biome.dev/linter/rules/no-shadow-restricted-names)
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "bar";
+    /// if (true) {
+    ///    const foo = "baz";
+    /// }
+    /// ```
+    ///
+    /// Variable declarations in functions can shadow variables in the outer scope:
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "bar";
+    /// const bar = function () {
+    ///     const foo = 10;
+    /// }
+    /// ```
+    ///
+    /// Function argument names can shadow variables in the outer scope:
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "bar";
+    /// function bar(foo) {
+    ///     foo = 10;
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const foo = "bar";
+    /// if (true) {
+    ///    const qux = "baz";
+    /// }
+    /// ```
+    ///
+    pub NoShadow {
+        version: "next",
+        name: "noShadow",
```

Comment:
What about `noShadowedVariables`? The other name is very generic

Reply:
From an end user's perspective, the rule doesn't just cover "variables". It covers all bindings, so functions, classes, imports, etc. I mean yes, technically they _are_ all variables in that you can read/write to them as such, but you wouldn't refer to them like that.

Plus, this is also a rare case where we can match the eslint rule name. :)